src/sparse/merkle_tree.rs

```diff
@@ -0,0 +1,385 @@
+                current_node = Node::create_node(&current_node, &side_node);
+            }
+            self.storage
+                .insert(&current_node.hash(), current_node.as_buffer())?;
+        }
+
+        self.root_node = current_node;
+
+        Ok(())
+    }
+
+    fn delete_with_path_set(
+        &'a mut self,
+        requested_leaf_node: &Node,
+        path_nodes: &[Node],
+        side_nodes: &[Node],
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        for node in path_nodes {
+            self.storage.remove(&node.hash())?;
+        }
+
+        let mut non_placeholder_reached = false;
+        let mut current_node = Node::create_placeholder();
+        let n = side_nodes.len();
+        for i in 0..n {
+            let side_node = &side_nodes[i];
+            if current_node.is_placeholder() {
+                if side_node.is_leaf() {
+                    current_node = side_node.clone();
+                    continue;
+                } else {
+                    non_placeholder_reached = true;
+                }
+            }
+
+            if !non_placeholder_reached && side_node.is_placeholder() {
+                continue;
+            } else if !non_placeholder_reached {
+                non_placeholder_reached = true;
+            }
+
+            let requested_leaf_key = requested_leaf_node.leaf_key();
+            if requested_leaf_key
+                .get_bit_at_index_from_msb(self.depth() - 1 - i)
+                .unwrap()
+                == 1
+            {
+                current_node = Node::create_node(&side_node, &current_node);
+            } else {
+                current_node = Node::create_node(&current_node, &side_node);
+            }
+            self.storage
+                .insert(&current_node.hash(), current_node.as_buffer())?;
+        }
+
+        self.root_node = current_node;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::common::{Bytes32, StorageError, StorageMap};
+    use crate::sparse::{Buffer, MerkleTree};
+    use hex;
+
+    #[test]
+    fn test_root_returns_empty_sum_with_no_leaves() {
+        let mut storage = StorageMap::<Bytes32, Buffer>::new();
+        let tree = MerkleTree::<StorageError>::new(&mut storage).unwrap();
+        let root = tree.root();
+        let expected_root = "0000000000000000000000000000000000000000000000000000000000000000";
+        assert_eq!(hex::encode(root), expected_root);
+    }
+
+    #[test]
+    fn test_update_1() {
+        let mut storage = StorageMap::<Bytes32, Buffer>::new();
+        let mut tree = MerkleTree::<StorageError>::new(&mut storage).unwrap();
+
+        for i in 0_u32..1 {
+            let key = i.to_be_bytes();
+            let data = "DATA".as_bytes();
+            let _ = tree.update(&key, data);
+        }
+
+        let root = tree.root();
+        let expected_root = "39f36a7cb4dfb1b46f03d044265df6a491dffc1034121bc1071a34ddce9bb14b";
+        assert_eq!(hex::encode(root), expected_root);
```

Comment:
It might be easier to manage potential changes to these hardcoded values in the future by using snapshot tests via [insta](https://insta.rs/). E.g.

```suggestion
        insta::assert_snapshot!(hex::encode(root))
```